service/debugger.go

```diff
@@ -0,0 +1,402 @@
+	if err != nil {
+		return fmt.Errorf("couldn't launch process: %s\n", err)
+	}
+
+	// Handle access to the process from the current thread.
+procLoop:
+	for {
+		select {
+		case f := <-d.processFuncs:
+			f(process)
+		case <-d.stop:
+			break procLoop
+		}
+	}
+
+	glog.Info("debugger stopping")
+	err = sys.PtraceDetach(process.Pid)
+	if err != nil {
+		glog.Errorf("couldn't detach from process: %s\n", err)
+	}
+	// TODO: could use the bool on the stop channel to drive this;
+	// or is that too clever?
+	err = process.Process.Kill()
+	if err != nil {
+		glog.Errorf("couldn't kill process: %s\n", err)
+	}
+	return nil
+}
+
+func (d *Debugger) State() (*api.DebuggerState, error) {
+	var state *api.DebuggerState
+
+	err := d.withProcess(func(p *proctl.DebuggedProcess) error {
+		var thread *api.Thread
+		th := p.CurrentThread
+		if th != nil {
+			thread = convertThread(th)
+		}
+
+		state = &api.DebuggerState{
+			// TODO: Running
+			// TODO: BreakPoint
+			Thread: thread,
+		}
+		return nil
+	})
+
+	return state, err
+}
+
+func (d *Debugger) CreateBreakPoint(requestedBp *api.BreakPoint) (*api.BreakPoint, error) {
+	var createdBp *api.BreakPoint
+	err := d.withProcess(func(p *proctl.DebuggedProcess) error {
+		loc := fmt.Sprintf("%s:%d", requestedBp.File, requestedBp.Line)
+		bp, breakError := p.BreakByLocation(loc)
+		if breakError != nil {
+			return breakError
+		}
+		createdBp = convertBreakPoint(bp)
+		return nil
+	})
+	return createdBp, err
+}
+
+func (d *Debugger) DeleteBreakPoint(requestedBp *api.BreakPoint) (*api.BreakPoint, error) {
+	var deletedBp *api.BreakPoint
+	err := d.withProcess(func(p *proctl.DebuggedProcess) error {
+		bp, err := p.Clear(requestedBp.Addr)
+		if err != nil {
+			return fmt.Errorf("Can't clear breakpoint @%x: %s\n", requestedBp.Addr, err)
+		}
+		deletedBp = convertBreakPoint(bp)
+		return nil
+	})
+	return deletedBp, err
+}
+
+func (d *Debugger) BreakPoints() []*api.BreakPoint {
+	bps := []*api.BreakPoint{}
+	d.withProcess(func(p *proctl.DebuggedProcess) error {
+		for _, bp := range p.HWBreakPoints {
+			if bp == nil {
+				continue
+			}
+			bps = append(bps, convertBreakPoint(bp))
+		}
+
+		for _, bp := range p.BreakPoints {
+			if bp.Temp {
+				continue
+			}
+			bps = append(bps, convertBreakPoint(bp))
+		}
+		return nil
+	})
+	return bps
+}
+
+func (d *Debugger) FindBreakPoint(id int) *api.BreakPoint {
+	for _, bp := range d.BreakPoints() {
+		if bp.ID == id {
+			return bp
+		}
+	}
+	return nil
+}
+
+func (d *Debugger) Threads() []*api.Thread {
+	threads := []*api.Thread{}
+	d.withProcess(func(p *proctl.DebuggedProcess) error {
+		for _, th := range p.Threads {
+			threads = append(threads, convertThread(th))
+		}
+		return nil
+	})
+	return threads
+}
+
+func (d *Debugger) FindThread(id int) *api.Thread {
+	for _, thread := range d.Threads() {
+		if thread.ID == id {
+			return thread
+		}
+	}
+	return nil
+}
+
+func (d *Debugger) ChangeThread(thread *api.Thread) (*api.DebuggerState, error) {
```

Comment:
What HTTP API do you propose for changing threads? Currently:

1. Send `GET /threads`, receive `[]Thread`.
2. Send `Thread` with `PUT /thread`, receive `DebuggerState`.